### ISSUE TITLE
Gradients do not exist with GradientTape()

### ISSUE THREAD

**Issue by sonnnguyen89** [tensorflow/probability#1560]
I was trying to build a BNN using `GradientTape()`, so I can have more access to the actual gradients, modify and alter the weights' update if needed.
`GradientTape()` seems not to work well at least with  `DenseVariational()` for `model1` with message:

_WARNING:tensorflow:Gradients do not exist for variables ['model__dense_variational/dense_v1/constant:0', 'model__dense_variational/dense_v2/constant:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss`argument?_

I don't think I have to compile the `model1` with loss function when using `GradientTape()`.

Code to re-produce as below, the code works fine with `model2` when I use `model.fit()`
tf.__version__ = 2.8.0
tfp.__version__ = 0.12.1

```
import numpy as np
import tensorflow as tf
import tensorflow_probability as tfp
from tensorflow_probability import layers as tpl
from tensorflow_probability import distributions as tfd
from tensorflow.keras import Model

negloglik = lambda y, rv_y: -rv_y.log_prob(y)

w0 = 0.125
b0 = 5.
x_range = [-20, 60]

def load_dataset(n=150, n_tst=150):
  np.random.seed(43)
  def s(x):
    g = (x - x_range[0]) / (x_range[1] - x_range[0])
    return 3 * (0.25 + g**2.)
  x = (x_range[1] - x_range[0]) * np.random.rand(n) + x_range[0]
  eps = np.random.randn(n) * s(x)
  y = (w0 * x * (1. + np.sin(x)) + b0) + eps
  x = x[..., np.newaxis]
  x_tst = np.linspace(*x_range, num=n_tst).astype(np.float32)
  x_tst = x_tst[..., np.newaxis]
  return y, x, x_tst

y, x, x_tst = load_dataset()


def posterior_mean_field(kernel_size, bias_size=0, dtype=None):
  n = kernel_size + bias_size
  c = np.log(np.expm1(1.))
  return tf.keras.Sequential([
      tfp.layers.VariableLayer(2 * n, dtype=dtype),
      tfp.layers.DistributionLambda(lambda t: tfd.Independent(
          tfd.Normal(loc=t[..., :n],
                     scale=1e-5 + tf.nn.softplus(c + t[..., n:])),
          reinterpreted_batch_ndims=1)),
  ])

def prior_trainable(kernel_size, bias_size=0, dtype=None):
  n = kernel_size + bias_size
  return tf.keras.Sequential([
      tfp.layers.VariableLayer(n, dtype=dtype),
      tfp.layers.DistributionLambda(lambda t: tfd.Independent(
          tfd.Normal(loc=t, scale=1),
          reinterpreted_batch_ndims=1)),
  ])

class Model_DenseVariational(Model):
    def __init__(self):
        super(Model_DenseVariational, self).__init__()
        self.dense_variational1 =  tpl.DenseVariational(10, posterior_mean_field, prior_trainable, activation='relu', 
                                  kl_weight=1/20, name='dense_v1')
        self.dense_variational2 =  tpl.DenseVariational(2, posterior_mean_field, prior_trainable, activation='relu', 
                                  kl_weight=1/20, name='dense_v2')
        self.DL =   tfp.layers.DistributionLambda(
                    lambda t: tfd.Normal(loc=t[..., :1],
                    scale=1e-3 + tf.math.softplus(0.05 * t[...,1:])))
    def call(self,input):
        x = self.dense_variational1(input)
        x = self.dense_variational2(x)
        return self.DL(x)

model1 = Model_DenseVariational()
optimizer = tf.keras.optimizers.Adam()
with tf.GradientTape() as tape:
    yhat = model1(x)
    loss = negloglik(y,yhat)
gradients = tape.gradient(loss, model1.trainable_variables)
optimizer.apply_gradients(zip(gradients, model1.trainable_variables))

model2 = Model_DenseVariational()
model2.compile(tf.keras.optimizers.Adam(), negloglik)
model2.fit(x, y)
```

**Comment by Frightera** [comment]
Even if you calculate `NLL`, training BNNs include 2 separate loss functions. One is NLL and the other is KL-Divergence between your variational posterior and prior.

 The KL-Term is added by `add.loss` and you need to consider it when implementing `GradientTape`:

```
optimizer = tf.keras.optimizers.Adam()
with tf.GradientTape() as tape:
    yhat = model1(x)
    loss = negloglik(y,yhat)
    loss = tf.reduce_mean(model1.losses) # You can access those loss values by model.losses
gradients = tape.gradient(loss, model1.trainable_variables)
optimizer.apply_gradients(zip(gradients, model1.trainable_variables))
```

**Comment by sonnnguyen89** [comment]
Thank you very much, adding the KL-Divergence loss fixed the zero gradients.
Just want to confirm, that the overall loss should be a combination of NLL and KL-Diverence, shouldn't it? Something like
```
    optimizer = tf.keras.optimizers.Adam()
    with tf.GradientTape() as tape:
        yhat = model(xTrain)
        loss = negloglik(labelTrain,yhat) - tf.reduce_mean(model1.losses)
    gradients = tape.gradient(loss, model1.trainable_variables)
    optimizer.apply_gradients(zip(gradients, model1.trainable_variables))))
```
Also, should it be `negloglik(labelTrain,yhat) - tf.reduce_mean(model.losses)` or `negloglik(labelTrain,yhat) + tf.reduce_mean(model.losses)`? The `tf.reduce_mean(model.losses)` returns both negative and positive values which confused me.

**Comment by Frightera** [comment]
Yes it should be `negloglik(labelTrain,yhat) + tf.reduce_mean(model.losses)`, I forgot to add them in my example. Since you don't use `kl_use_exact` in `DenseVariational`, those values are approximate of the KL loss per layer.

**Comment by sonnnguyen89** [comment]
> Yes it should be `negloglik(labelTrain,yhat) + tf.reduce_mean(model.losses)`, I forgot to add them in my example. Since you don't use `kl_use_exact` in `DenseVariational`, those values are approximate of the KL loss per layer.

Thank you very much, I'm closing the issue.
